packages/opentelemetry-core/src/default_span.ts

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Span,
+  SpanContext,
+  Status,
+  Attributes,
+  SpanKind,
+} from '@opentelemetry/types';
+
+/** This is a no-op implementation of Span. */
+export class DefaultSpan implements Span {
+  protected name = 'DefaultSpan'; // default
+  protected spanId = '';
```

Comment:
Should `traceId`, `spanId` and `parentSpanId` be set on the `SpanContext` instead? Otherwise how can they be propagated?

Reply:
I thought will use `context() / _context()` method to construct the `SpanContext`;
Something like,
```
protected _context(): SpanContext {
    return { traceId: this.traceId, spanId: this.spanId, ... };
}
```
WDYT?

Reply:
Why not assign it in the constructor?

Reply:
Sure, it is a good idea to do in the constructor. But looks like this is only applicable in case `traceId` and `spanId` right? `SpanContext ` does not have `parentSpanId` or `name` or `kind` properties.

Reply:
Why not? 🤔In general I would put as much of the state as possible on the span context since you can have a span context without a span but the opposite is not true.

Reply:
AFAIK `SpanContext` must be **immutable** (can't use them for properties like `name`, `status`, `attributes`, `links` etc) and should contain only **propagating (and serializable) information.**

@danielkhan @draffensperger @bg451 @hekike any comment?

Reply:
This is very limiting. The whole point of `SpanContext` (at least in OpenTracing) is to allow vendors to attach all sorts of metadata, with some metadata being included by default. It also makes it much easier to implement things like samplers for example since it publicly exposes the metadata.

Reply:
@SergeyKanzhelev and @bogdandrutu we would like to know your comment on this.

Reply:
I like the idea of storing the whole `SpanContext` in the span instead of destructing. @rochdev the vendor-specific metadata would live in the [this.spanContext.traceState](https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-types/src/trace/span_context.ts#L62) right?

Reply:
Possibly, but then `TraceState` should always be an object and not a string, which may be an issue. The fact that it's a string in the spec probably means it's the string representation of some other field(s).

Reply:
I see what you are saying, https://www.w3.org/TR/trace-context/#tracestate-field
1. One idea could be that we store the `parsedTraceState` as key-value pairs on the span and we serialize them for propagation. But it would come with a trade-off around unnecessary de/serializations when users don't want to read/write them. 
2. An alternative solution could be to expose it via APIs like `addTraceStateMember(key, value)`, `removeTraceStateMember(key)`, parseTraceState(): { [string]: string }` and parse it only when someone uses the API. (naming is hard, this is just an example :))

Reply:
In general, how is `traceState` supposed to be serialized/deserialized? I think this is important as it needs to support this API, whatever it ends up being. My personal opinion is that the `SpanContext` is way too early a place to hold the `traceState` as a string which, which only needs to be serialized at propagation time.

Reply:
We cannot accept any random object because then you have problems with encoding/decoding. So it has to be encoded as a string.

SpanContext MUST be immutable in order to avoid complicated things when this object is shared between threads.

Reply:
> We cannot accept any random object because then you have problems with encoding/decoding. So it has to be encoded as a string.

This doesn't seem like a real issue, especially for JavaScript. Encoding/decoding is easy.

> SpanContext MUST be immutable in order to avoid complicated things when this object is shared between threads.

This concern doesn't apply to JavaScript since it runs on a single thread. If `SpanContext` is immutable, then to me it means it's an object that must happen way later in the pipeline. Otherwise there must be an alternative that replicates the behavior of `SpanContext` from OpenTracing. Having this alternative would however make the OpenTelemetry `SpanContext` basically useless, so I don't see why not just use `SpanContext` for this purpose.